File: java/data_structures_examples/sparse_table.cpp
#include <bits/stdc++.h>
using namespace std;

// https://leetcode.com/problems/largest-rectangle-in-histogram/description/
class Solution
{
public:
    int largestRectangleArea(vector<int> &heights)
    {
        int n = heights.size();
        int lg = (int)log2(heights.size());
        vector<vector<int>> st(lg + 1, vector<int>(n));

        for (int l = 0; l <= lg; l++)
        {
            for (int i = 0; i + (1 << l) <= n; i++)
            {
                if (l == 0)
                {
                    st[l][i] = i;
                }
                else
                {
                    int li = st[l - 1][i], ri = st[l - 1][i + (1 << (l - 1))];
                    st[l][i] = heights[li] <= heights[ri] ? li : ri;
                }
            }
        }

        auto ff = [&](int l, int r)
        {
            int lg = log2(r - l + 1);
            int li = st[lg][l], ri = st[lg][r - (1 << lg) + 1];
            return heights[li] <= heights[ri] ? li : ri;
        };

        function<int(int, int)> dc = [&](int l, int r)
        {
            if (l > r)
            {
                return 0;
            }
            int minindex = ff(l, r);
            return max(heights[minindex] * (r - l + 1), max(dc(l, minindex - 1), dc(minindex + 1, r)));
        };

        return dc(0, n - 1);
    }
};

File: java/data_structures_examples/sparse_table.py
# https://leetcode.com/problems/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/description/
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        n = len(nums)

        mn = [[0] * n for _ in range(int(math.log2(n)) + 1)]
        mx = [[0] * n for _ in range(int(math.log2(n)) + 1)]

        for i in range(int(math.log2(n)) + 1):
            for j in range(n - (1 << i) + 1):
                if not i:
                    mn[i][j] = nums[j]
                    mx[i][j] = nums[j]
                else:
                    mn[i][j] = min(mn[i-1][j], mn[i-1][j + (1 << (i - 1))])
                    mx[i][j] = max(mx[i-1][j], mx[i-1][j + (1 << (i - 1))])

        def get(l, r, min_or_max, st):
            ll = r - l + 1
            lg2 = int(math.log2(ll))
            return min_or_max(st[lg2][l], st[lg2][r - (1 << lg2) + 1])

        ans = 0
        s = 0
        for e in range(n):
            while abs(get(s, e, max, mx) - get(s, e, min, mn)) > limit:
                s += 1
            ans = max(ans, e - s + 1)
        return ans